pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -795,16 +799,28 @@ func (p *DNSProxy) LookupEndpointByIP(ip netip.Addr) (endpoint *endpoint.Endpoin
 
 // UpdateAllowed sets newRules for endpointID and destPort. It compiles the DNS
 // rules into regexes that are then used in CheckAllowed.
-func (p *DNSProxy) UpdateAllowed(endpointID uint64, destPortProto restore.PortProto, newRules policy.L7DataMap) error {
+func (p *DNSProxy) UpdateAllowed(endpointID uint64, destPortProto restore.PortProto, newRules policy.L7DataMap) (revert.RevertFunc, error) {
 	p.Lock()
 	defer p.Unlock()
 
 	err := p.allowed.setPortRulesForID(p.cache, endpointID, destPortProto, newRules)
-	if err == nil {
-		// Rules were updated based on policy, remove restored rules
-		p.removeRestoredRulesLocked(endpointID)
+	if err != nil {
+		return nil, err
 	}
-	return err
+
+	// Rules were updated based on policy, remove restored rules
+	p.removeRestoredRulesLocked(endpointID)
+
+	// Get current rules for reverting
+	oldRules := p.currentRules
+	p.currentRules = newRules
+
+	revert := func() error {
+		p.Lock()
+		defer p.Unlock()
+		return p.allowed.setPortRulesForID(p.cache, endpointID, destPortProto, oldRules)
+	}
+	return revert, nil
```

Comment:
(Trying to understand the flow of new rules here)
Are the `newRules` not specific to that endpoint ? If so, the how is `proxy.currentRules` going to differentiate if there two consecutive updates ?
T0 : Endpoint 1 will get the proxy lock to update the `currentRules` as endpoint 1 rules. 
T1: Endpoint 2 also now updates the proxy `current rules` as proxy is a singleton, but the revert function for endpoint 2 will have the old rules(set by endpoint 1) which is incorrect ?